functions/Set-DbaCompression.ps1

```diff
@@ -0,0 +1,508 @@
+                                    SET     AlreadyProcessed = 1
+                                    WHERE   PK = @PK;
+
+                                    SET @CompressedCount = @CompressedCount
+                                    + 1;
+                                 END
+                              ELSE
+                                 BEGIN
+                                       PRINT 'Max runtime reached. Some compression performed. Exiting...';
+                                       BREAK
+                                 END
+
+                              FETCH cCompress INTO @Schema, @TableName, @Partition, @IndexName,
+                                    @IndexType, @CompressionTypeRecommendation, @PK;
+                        END
+
+                  CLOSE cCompress;
+                  DEALLOCATE cCompress;
+
+                  SELECT 
+                   PK				
+                  ,DBName = DB_Name()
+				  ,[Schema] 
+				  ,[TableName] 
+				  ,[IndexName] 
+				  ,[Partition] 
+				  ,[IndexID] 
+				  ,[IndexType] 
+				  ,[PercentScan] 
+				  ,[PercentUpdate] 
+				  ,[RowEstimatePercentOriginal] 
+				  ,[PageEstimatePercentOriginal]
+				  ,[CompressionTypeRecommendation] 
+				  ,SizeCurrentKB = [SizeCurrent]
+				  ,SizeRequestedKB = [SizeRequested]
+                  ,PercentCompression
+				  ,AlreadyProcessed
+				  FROM ##setdbacompression
+                  WHERE AlreadyProcessed=1;
+
+                  IF OBJECT_ID('tempdb..##setdbacompression' , 'U') IS NOT NULL
+                  DROP TABLE ##setdbacompression
+				
+                  IF OBJECT_ID('tempdb..##tmpEstimateRow' , 'U') IS NOT NULL
+                  DROP TABLE ##tmpEstimateRow
+
+                  IF OBJECT_ID('tempdb..##tmpEstimatePage' , 'U') IS NOT NULL
+                  DROP TABLE ##tmpEstimatePage;"
+	}
+	
+	process {
+		
+		foreach ($instance in $SqlInstance) {
+			try {
+				Write-Message -Level VeryVerbose -Message "Connecting to $instance" -Target $instance
+				$server = Connect-SqlInstance -SqlInstance $instance -SqlCredential $SourceSqlCredential -MinimumVersion 10
+			}
+			catch {
+				Stop-Function -Message "Failed to process Instance $Instance" -ErrorRecord $_ -Target $instance -Continue
+			}
+			
+            $Server.ConnectionContext.StatementTimeout = 0
+                
+                if ($Server.EngineEdition -eq 'Standard' -and $Server.VersionMajor -lt '13')
+                    {
+                    Stop-Function -Message "Only SQL Server Enterprise Edition supports compression on $Server" -Target $Server -Continue
+                    }
+			#If IncludeSystemDBs is true, include systemdbs
+			#look at all databases, online/offline/accessible/inaccessible and tell user if a db can't be queried.
+			try {
+				if ($Database) {
+					$dbs = $server.Databases | Where-Object Name -In $Database
+				}
+				elseif ($IncludeSystemDBs) {
+					$dbs = $server.Databases | Where-Object Status -eq 'Normal'
```

Comment:
This will end up including databases that are on a secondary replica (AG or mirroring) but not readable. It is best to check for `IsAccessible` as that guarantees you can query the database, the status does not.